requests-for-comments/2022-11-08-supported-languages.md

```diff
@@ -0,0 +1,169 @@
+
+Candidate Languages:
+
+- [Golang](https://go.dev/) (big surprise)
+- [Rust](https://www.rust-lang.org/) 🦀
+- [OCaml](https://ocaml.org/)
+- [Elixir](https://elixir-lang.org/) (dynamically typed)
+- [Scala](https://www.scala-lang.org/)
+- [Java](<https://en.wikipedia.org/wiki/Java_(programming_language)>)
+
+## Meet the eligible languages
+
+### Golang
+
+The good:
+
+- Extremely easy to learn. Most engineers can learn and be productive in about a day.
+- Built to remove most contentious parts of development
+- Designed to make engineering easier
+- Built for network services and concurrency
+- Great standard lib
+- Super fast compile times
+- Light on memory
+- Used by plenty of organizations big and small
+- We have plenty of Gophers here.
+- A simple binary as a deliverable.
+
+The bad:
+
+- Most people would not say the language sparks joy (I disagree)
+- Not terribly expressive
+- Verbose, but easy to read
+
+### Rust
+
+The good:
+
+- Very hot right now
+- Very good for when you need to hyper optimize something but would like to avoid C or C++
+- Extraordinarily expressive and fun to program
+- Extremely performant and light on memory
+- Used by plenty of organizations big and small
+- We have a few Crustaceans here!
+- A simple binary as a deliverable.
+
+The bad:
+
+- Slow compile times
+- Harder to read because of how expressive
+- Slower ramp up time to become proficient in as an Engineer
+
+### OCaml
+
+> OCaml is a general-purpose, industrial-strength programming language with an emphasis on expressiveness and safety.
+
+You may not be familiar with this one but it is _very_ safe and performant. Companies that have huge amounts of money on the line who are risk adverse to defects and slowdowns, like the HFT firm Jane Street use OCaml because it is unforgiving in how type safe it is.
+
+> OCaml’s powerful type system means more bugs are caught at compile time, and large, complex codebases are easier to maintain. This makes it a good language for running critical code. At the same time, sophisticated inference makes the type system unobtrusive, creating a smooth developer experience.
+
+The good:
+
+- Garbage collected
+- Algebraic data types
+- Pattern matching
+- Type inference
+- Immutable
+- Static Type-checking
+- First-class functions
+- Parametric polymorphism
+- Used by serious programming shops
+
+The bad:
+
+- Not terribly popular relative to the others
+- Can be considered relatively academic
+- No one experienced here
+
+### Elixir
+
+Oh, Erlang <3
+
+The good:
+
+- Built for streaming data
+- The cockroach of runtimes.
+- Automatic function level clustering
+- Hot reloading of functions
+- Compiled
+- Relatively functional
+
+The bad:
+
+- Dynamically typed
+- Runs in a VM
+
+### Scala / Java
+
+Grouping these together because they really are converging. Really this is any language on the JVM.
+
+The JVM is truly a wonderful runtime. It's very fast. The hotspot detection + Just In Time Compiling is _super_ impressive.
```

Comment:
I saw a talk by the folk at a German car sales company who were shipping an ML model in Java. They had to write tooling to hit all of the code paths they cared about during boot because JIT was a problem for them.

That said there are a bunch of alternative VMs that are focussed on speed because of serverless-style loads...